### PR TITLE
Initial HIP binding build support using headers from standard HIP install

### PR DESCRIPTION
**Pull request by sonals**
#### Problem solved by the commit
Use standard HIP install for HIP runtime header files

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
HIP binding build support

#### How problem was solved, alternative solutions (if any) and why they were rejected
Install HIP in xrtdeps.h and then leverage hipconfig utility to find HIP include directories 

#### Risks (if any) associated the changes in the commit
None; only enabled when build is run with -hip switch and is limited to Linux

#### What has been tested and how, request additional testing if necessary
Compile tested on Ubuntu 23.10

#### Documentation impact (if any)
None

### CODE REVIEW

**Comment by sonals** [comment]
retest this please

File: tests/hip/vadd/VectorAdd/common.h [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+class HipDevice {
+private:
+    hipDevice_t mDevice;
+    int mIndex;
+    std::map<std::string, hipModule_t> mModuleTable;
+
+public:
+    HipDevice(int index = 0) : mIndex(index) {
+        hipCheck(hipDeviceGet(&mDevice, index));
+    }
```
Comment (stsoe) [inline tests/hip/vadd/VectorAdd/common.h:89]: Avoid constructor body for simple initializations.  Re-factor
```suggestion
    hip_device(int index = 0) 
      : m_index{index}
      , m_device{hip_check(hip_get_device(m_index)}
    {}
```
With 
```
static hipDevice_t
hip_get_device(int index)
{
  hipDevice_t device;
  hip_check(hipGetDevice(&device, index));
  return devlce;
}
```

File: tests/hip/vadd/VectorAdd/common.h [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    T *&get() {
+        return _buffer;
+    }
+
+};
+
+class HipDevice {
+private:
+    hipDevice_t mDevice;
+    int mIndex;
```
Comment (stsoe) [inline tests/hip/vadd/VectorAdd/common.h:83]: Swap
```suggestion
  int m_index;
  hipDevice_t m_device;
```

File: tests/hip/vadd/VectorAdd/common.h [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    std::map<std::string, hipModule_t> mModuleTable;
+
+public:
+    HipDevice(int index = 0) : mIndex(index) {
+        hipCheck(hipDeviceGet(&mDevice, index));
+    }
+
+    virtual ~HipDevice() {
```
Comment (stsoe) [inline tests/hip/vadd/VectorAdd/common.h:91]: Why virtual?

**Review by stsoe — CHANGES_REQUESTED** [review]
Many style comments.   This will be monkey-see-monkey-do, do we really want that? Or should we spend the time formatting this newly added code to perfection (whatever perfection is)?

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

File: tests/hip/vadd/VectorAdd/main-stream.cpp [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+#define NOP_FILENAME "nop.co"
+#define NOP_KERNELNAME "mynop"
+
+namespace {
+
+static const int LEN = 0x100000;
+static const int SIZE = LEN * sizeof(float);
+static const int THREADS_PER_BLOCK_X = 32;
+static const int LOOP = 1000;
```
[the remark this answers is not in the document]
Comment (sonals) [reply tests/hip/vadd/VectorAdd/main-stream.cpp]: Updated to use constexpr and lowercase

File: tests/hip/vadd/VectorAdd/common.h [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+#include <chrono>
+#include <cstring>
+#include <iostream>
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <system_error>
+
+#include <boost/uuid/uuid.hpp>
+#include <boost/uuid/uuid_io.hpp>
```
[the remark this answers is not in the document]
Comment (sonals) [reply tests/hip/vadd/VectorAdd/common.h]: Using boost essentially for OS independent way to print UUID to ostream. Does XRT UUID support printing?

File: tests/hip/vadd/VectorAdd/common.h [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+#include "hip/hip_runtime_api.h"
+
+class HIPError : public std::system_error
```
[the remark this answers is not in the document]
Comment (sonals) [reply tests/hip/vadd/VectorAdd/common.h]: Updated to match the coding style.

File: tests/hip/vadd/VectorAdd/main-stream.cpp [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+/* Copyright (C) 2022-2023 Advanced Micro Devices, Inc. */
+/* Based on https://github.com/ROCm-Developer-Tools/HIP-Examples/tree/master/vectorAdd */
+
+#include <cstring>
+#include <algorithm>
+#include <iostream>
+#include <memory>
+
+#include <boost/uuid/uuid.hpp>
+#include <boost/uuid/uuid_io.hpp>
```
[the remark this answers is not in the document]
Comment (sonals) [reply tests/hip/vadd/VectorAdd/main-stream.cpp]: Obsoleted this file

File: tests/hip/vadd/VectorAdd/main-stream.cpp [hunk 8]
```diff
@@ -0,0 +1,200 @@
+/* SPDX-License-Identifier: Apache License 2.0 */
+/* Copyright (c) 2021-2022 Xilinx, Inc. All rights reserved */
+/* Copyright (C) 2022-2023 Advanced Micro Devices, Inc. */
+/* Based on https://github.com/ROCm-Developer-Tools/HIP-Examples/tree/master/vectorAdd */
```
[the remark this answers is not in the document]
Comment (sonals) [reply tests/hip/vadd/VectorAdd/main-stream.cpp]: Obsoleted this file. Updated main.cpp to match the style suggested.

File: tests/hip/vadd/VectorAdd/common.h [hunk 9]
```diff
@@ -0,0 +1,126 @@
+/* SPDX-License-Identifier: Apache License 2.0 */
+/* Copyright (C) 2023 Advanced Micro Devices, Inc. */
```
[the remark this answers is not in the document]
Comment (sonals) [reply tests/hip/vadd/VectorAdd/common.h]: Updated to match the style suggested.

File: tests/hip/vadd/VectorAdd/Makefile [hunk 10]
```diff
@@ -0,0 +1,54 @@
+# SPDX-License-Identifier: Apache License 2.0 #
+# Copyright (c) 2021-2022 Xilinx, Inc. All rights reserved #
+# Copyright (C) 2022-2023 Advanced Micro Devices, Inc. #
```
[the remark this answers is not in the document]
Comment (sonals) [reply tests/hip/vadd/VectorAdd/Makefile]: Obsoleted this file.

File: tests/hip/vadd/VectorAdd/common.h (hunk 2 above)
Comment (sonals) [reply tests/hip/vadd/VectorAdd/common.h]: Updated to match the coding style.

File: tests/hip/vadd/VectorAdd/common.h (hunk 3 above)
Comment (sonals) [reply tests/hip/vadd/VectorAdd/common.h]: Removed

**Comment by sonals** [comment]
retest this please

File: tests/hip/vadd/VectorAdd/common.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (sonals) [reply tests/hip/vadd/VectorAdd/common.h]: Dropped boost completely and copied code from XRT uuid.h for printing.

**Review by stsoe — APPROVED** [review]
Looks good.   So by default hip will only be supported on Ubuntu 23+ or fedora > 38, not of which are built by CI?  And anyway hip build has to be enabled explicitly through build switch, so  assume this is all good enough for now.

File: tests/hip/vadd/main.cpp [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+mainworker() {
+
+  std::cout << "---------------------------------------------------------------------------------\n";
+  hip_test_device hdevice;
+  hdevice.show_info(std::cout);
+
+  hipFunction_t function = hdevice.get_function(kernel_filename, kernel_name);
+  hipFunction_t nopfunction = hdevice.get_function(nop_kernel_filename, nop_kernel_name);
+
+  std::unique_ptr<float[]> host_a(new float[vector_length]);
```
Comment (keryell) [inline tests/hip/vadd/main.cpp:79]: Why not just `std::vector`?

**Pull request merged**